056-zk-kraft-migration.md

```diff
@@ -0,0 +1,258 @@
+
+#### Phase 1: Provisioning the KRaft controller quorum
+
+Starting from a `NoMigration` (or null) state (with the `strimzi.io/kraft-migration-state` annotation missing), the user applies the `strimzi.io/kraft-migration: pre-migration` on the `Kafka` custom resource to make the operator aware of the migration intention.
+The FSM moves to a `WaitControllersInMigration` state where the operator expects the user to create the KRaft controllers pool.
+The user creates the `KafkaNodePool` custom resource containing the definition of the KRaft controllers which are going to replace the ZooKeeper nodes.
+The operator creates and configures the controllers accordingly with the ZooKeeper migration flag and the connection to the ZooKeeper ensemble. 
+
+```shell
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+
+# ZooKeeper client configuration
+zookeeper.connect=<connection-to-zookeeper>
+```
+
+> NOTE: without an annotation on the `Kafka` custom resource to differentiate the operator behavior, the operator itself just creates the controllers in the pool and restarts the Kafka brokers to use them. ZooKeeper is not used anymore but no proper migration takes place (from ZooKeeper to KRaft metadata). Because of KRaft and related migration from ZooKeeper not supported yet, this can drive to the following [issue](https://github.com/strimzi/strimzi-kafka-operator/issues/9108).
+
+The operator has to wait for the KRaft controllers being ready, the quorum to be formed and leader elected, then the migration process moves to the state of waiting for brokers to register.
+
+The FSM moves to a `WaitBrokersInMigration` state where the operator expects the user to ask for running the actual metadata migration.
+It means the operator has to enable the ZooKeeper migration on the brokers pool in order to actually start the migration (Phase 2).
+
+From this state it's still possible to not move forward with the migration.
+The user can delete the `KafkaNodePool` custom resource containing the definition of the KRaft controllers and remove the `strimzi.io/kraft-migration` annotation from the `Kafka` custom resource.
+The FSM moves back to the `NoMigration` state (or null) actually exiting from the migration process and the operator removes the `strimzi.io/kraft-migration-state` annotation from the `Kafka` custom resource.
+There is nothing to do with the brokers which are still configured to work with the ZooKeeper ensemble only.
+
+#### Phase 2: Enabling the migration on the brokers
+
+In the `WaitBrokersInMigration` state, the operator expects the user to ask for running the actual metadata migration.
+In order to start the actual migration, the brokers need to be connected to the KRaft controllers and have the ZooKeeper migration enabled.
+The user annotates the `Kafka` custom resource running the current ZooKeeper based brokers with `strimzi.io/kraft-migration: migration`.
```

Comment:
How does the user find out?

Reply:
It should check the migration status stored "somewhere" in the `Kafka` custom resource (see discussion using annotation state vs status field).

Reply:
After latest change, there is the `Kafka.status.migrationState` the user has to check.